docs/developer/style_guide.md

```diff
@@ -0,0 +1,579 @@
+maintain in the future. Fortran 2003 has been chosen as the language that all
+the code will be written in, as it has many modern features which enable the
+writing of clean, modular code without compromising efficiency and portability.
+
+In this document, there is a [template for writing modules](#template), and for
+writing subprograms within a module. These templates are designed to enforce
+good programming practice and documentation of routine function and
+side-effects.  It is important that all routines be written with these
+templates. The general guidelines about style addressed here, the use of module
+templates, the completeness of the documentation, etc. all help to make writing
+new code and maintaining the existing code a simple task.
+
+It is also the programmer's responsibility to ensure that each routine is coded
+as efficiently as possible. Whilst it is tempting to use some of the new
+built-in features of Fortran 2003, experience has shown that discretion is
+required in their usage so as not to compromise efficiency. Therefore,
+guidelines on the use of such features has also been included.
+
+
+## Writing to the Specification Document
+
+
+### "It does exactly what it says on the tin"
+
+Obviously, it is of paramount importance to make sure that any routine that you
+are responsible for does exactly what the specification says that it should do,
+particularly with regard to exit conditions, side-effects, etc.
+
+Routines and variables that are covered by the specification are `public` whilst
+other routines and variables are `private`. The specification says nothing about
+the rôle or meaning of private routines, although of course, they should be
+documented within the Implementation Guide. Private routines and variables are
+only accessible within the module in which they occur.
+
+
+### Do not assume anything that is not written in the specification document
+
+If the specification does not say that a particular routine will preserve the
+input values on exit, then you cannot presume that whoever coded it will do
+likewise! Even if that routine was written by you...
+
+Similarly, it may seem obvious that a particular variable has a particular
+meaning, but if it is not defined in the specification, then it cannot be
+presumed. Future revisions of the code may change non-specified routines
+freely. If all code sticks to the specification at all times, then maintaining
+the code is much simpler.
+
+Note that the specification does not say anything about private routines or
+variables in any module. For this reason, great care must be taken when using
+such routines. Use the Implementation Guide as a reference for what such
+routines/variables do, and if you make any modifications, you must update the
+Implementation Guide and note the changes within the source code as well. The
+subprogram template is there to remind you to document all side-effects of each
+routine!
```

Comment:
Do we still want to refer to specification given the spec has not been updated for a while?

Reply:
Yes, I think this is OK. One day we may move the spec to a more modern format ...